di/container/_solving.py

```diff
@@ -136,6 +170,16 @@ def solve(
 
     dag, parents = build_dag(dependency, binds)
 
+    def get_scope(dep: "DependantBase[Any]") -> Scope:
+        if dep.scope is not None:
+            return dep.scope
+        if None in scopes:
+            return dep.scope
+        path = get_path(dep, parents)
+        dep_scopes = [d.scope for d in path]
+        # dep is the last dependency in path
+        return next(iter(reversed(resolve_unset_scopes(dep_scopes, scopes))))
```

Comment:
Not sure if a closure is the best way to do this or we should move this into `build_tasks`